model_zoo/census_model_sqlflow/wide_and_deep/census_wide_and_deep.sql

```diff
@@ -0,0 +1,12 @@
+SELECT *
+FROM census_income
+TO TRAIN WideAndDeepClassifier
+COLUMN (
+    SET GROUP(APPLY_VOCAB(workclass), BUCKETIZE(capital_gain, bucket_num=5), BUCKETIZE(capital_loss, bucket_num=5), BUCKTIZE(hours_per_week, bucket_num=6)) AS group_1,
```

Comment:
SET means value assignment here -- it is a very different philosophy from the functional programming style we have been used in the design of SQLFlow syntax. Can we don't use SET here?

@typhoonzero @shendiaomo Do you have suggestions here?

Reply:
`SET` can be used in three ways:

1. user-defined variables
2. setting variables in procedures
3. setting system variables like charset.

see: https://dev.mysql.com/doc/refman/5.7/en/set-variable.html, https://dev.mysql.com/doc/refman/5.7/en/user-variables.html

In this case, I assume that you want to use `SET` to set a user-defined variable, and the variable is a list of feature columns. For a user-defined variable, it can be set in one SQL statement and referred to in another SQL statement. Then it should look like: 

```sql
SET group_1 = GROUP(APPLY_VOCAB(workclass), ...)
SELECT * .... COLUMN EMBEDDING(group1) ...
```

Then the statement `SET group_1 = GROUP(APPLY_VOCAB(workclass), ...)` will be passed to the third-party parser and it will report parsing error. And, implement the whole `SET` grammar parsing inside `COLUMN` does not make sense either.

I recommend the grammar should look like:

```sql
COLUMN
    EMBEDDING(
        CONCAT(
            CATEGORY_ID_HASH(col1, 64),
            CATEGORY_ID_VOCAB(col2),
            col3,
        )
    ) FOR linear_feature_columns
COLUMN .... FOR dnn_feature_columns
```

Reply:
NOTE, the feature column `CATEGORY_ID_VOCAB` can be omitted since the vocabulary of the column can be done by feature derivation.

Reply:
> NOTE, the feature column `CATEGORY_ID_VOCAB` can be omitted since the vocabulary of the column can be done by feature derivation.

I curious about how we calculate the vocabulary list inside feature derivation?

I believe we need do a statistic work using SQL to calculate the vocabulary just like [this](https://github.com/sql-machine-learning/elasticdl/issues/1667#issuecomment-580044742)

Reply:
> ```sql
> COLUMN
>     EMBEDDING(
>         CONCAT(
>             CATEGORY_ID_HASH(col1, 64),
>             CATEGORY_ID_VOCAB(col2),
>             col3,
>         )
>     ) FOR linear_feature_columns
> ```

If we write two COLUMN clause, I'm afraid the following clause will be written twice. It will be user unfriendly if there are hundreds of columns which is a common in a search&ranking scenario:

```
    SET GROUP(APPLY_VOCAB(workclass), BUCKETIZE(capital_gain, bucket_num=5), BUCKETIZE(capital_loss, bucket_num=5), BUCKTIZE(hours_per_week, bucket_num=6)) AS group_1,
    SET GROUP(HASH(education), HASH(occupation), APPLY_VOCAB(martial_status), APPLY_VOCAB(relationship)) AS group_2,
    SET GROUP(BUCKETIZE(age, bucket_num=5), HASH(native_country), APPLY_VOCAB(race), APPLY_VOCAB(sex)) AS group_3,
```

Reply:
Maybe we use a nested COLUMN syntax to implement complicated transformation like
```python
SELECT *
FROM census_income
TO TRAIN WideAndDeepClassifier
COLUMNS (
    [EMBEDDING(group1, 1), EMBEDDING(group2, 1)] AS wide_embeddings
    [EMBEDDING(group1, 8), EMBEDDING(group2, 8), EMBEDDING(group3, 8)] AS deep_embeddings
    FROM(
        CONCAT(APPLY_VOCAB(workclass), BUCKETIZE(capital_gain, bucket_num=5), BUCKETIZE(capital_loss, bucket_num=5), BUCKTIZE(hours_per_week, bucket_num=6)) AS group_1,
        CONCAT(HASH(education), HASH(occupation), APPLY_VOCAB(martial_status), APPLY_VOCAB(relationship)) AS group_2,
        CONCAT(BUCKETIZE(age, bucket_num=5), HASH(native_country), APPLY_VOCAB(race), APPLY_VOCAB(sex)) AS group_3,
    )
)
LABEL label
```
Then, we can parse the COLUMN expression in order.

Reply:
We also can place embedding in the model definition and users don't need to write Embedding.
```sql
SELECT *
FROM census_income
TO TRAIN WideAndDeepClassifier
COLUMNS (
    [group1, group2] AS wide
    [group1, group2, group3] AS deep
    FROM(
        CONCAT(APPLY_VOCAB(workclass), BUCKETIZE(capital_gain, bucket_num=5), BUCKETIZE(capital_loss, bucket_num=5), BUCKTIZE(hours_per_week, bucket_num=6)) AS group_1,
        CONCAT(HASH(education), HASH(occupation), APPLY_VOCAB(martial_status), APPLY_VOCAB(relationship)) AS group_2,
        CONCAT(BUCKETIZE(age, bucket_num=5), HASH(native_country), APPLY_VOCAB(race), APPLY_VOCAB(sex)) AS group_3,
    )
)
LABEL label
```

Reply:
After discussion with @typhoonzero @workingloong @shendiaomo , We will rename `GROUP` to `CONCAT`.

Reply:
Why there are two COLUMN clauses in the third option? So far I'd vote for the second option.

Reply:
After more discussion with @wangkuiyi , the third option is more inline with the SQL style. We will choose this design.

```SQL
SELECT *
FROM census_income
TO TRAIN WideAndDeepClassifier
COLUMN
    EMBEDDING(CONCAT(APPLY_VOCAB(workclass), BUCKETIZE(capital_gain, bucket_num=5), BUCKETIZE(capital_loss, bucket_num=5), BUCKTIZE(hours_per_week, bucket_num=6)) AS group_1, 8),
    EMBEDDING(CONCAT(HASH(education), HASH(occupation), APPLY_VOCAB(martial_status), APPLY_VOCAB(relationship)) AS group_2, 8),
    EMBEDDING(CONCAT(BUCKETIZE(age, bucket_num=5), HASH(native_country), APPLY_VOCAB(race), APPLY_VOCAB(sex)) AS group_3, 8)
    FOR deep_embeddings
COLUMN
    EMBEDDING(group1, 1),
    EMBEDDING(group2, 1)
    FOR wide_embeddings
LABEL label
```